account/rethinkdb/store.go

```diff
@@ -0,0 +1,279 @@
+package rethinkdb
+
+import (
+	"fmt"
+
+	rdb "github.com/dancannon/gorethink"
+	"github.com/mitchellh/mapstructure"
+
+	"github.com/asaskevich/govalidator"
+	"github.com/go-errors/errors"
+	"github.com/ory-am/common/pkg"
+	"github.com/ory-am/hydra/account"
+	"github.com/ory-am/hydra/hash"
+	"github.com/pborman/uuid"
+)
+
+const accountsTable = "hydra_account"
+
+type Store struct {
+	hasher  hash.Hasher
+	session *rdb.Session
+}
+
+func New(h hash.Hasher, session *rdb.Session) *Store {
+	return &Store{h, session}
+}
+
+func (s *Store) CreateTables() error {
+	exists, err := s.tableExists(accountsTable)
+	if err == nil && !exists {
+		_, err := rdb.TableCreate(accountsTable).RunWrite(s.session)
+		if err != nil {
+			fmt.Println(err)
+		}
+	}
+	return nil
+}
+
+// TableExists : check if table(s) exists in database
+func (s *Store) tableExists(table string) (bool, error) {
+
+	res, err := rdb.TableList().Run(s.session)
+	if err != nil {
+		return false, err
+	}
+	defer res.Close()
+
+	if res.IsNil() {
+		return false, nil
+	}
+
+	var tableDB string
+	for res.Next(&tableDB) {
+		if table == tableDB {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+func validate(r interface{}) error {
+	if v, err := govalidator.ValidateStruct(r); !v {
+		return pkg.ErrInvalidPayload
+	} else if err != nil {
+		return pkg.ErrInvalidPayload
+	}
+	return nil
+}
+
+func (s *Store) Create(r account.CreateAccountRequest) (account.Account, error) {
+	var err error
+
+	if r.ID == "" {
+		r.ID = uuid.New()
+	}
+
+	if r.Data == "" {
+		r.Data = "{}"
+	}
+
+	if err := validate(r); err != nil {
+		return nil, err
+	}
+
+	// Hash the password
+	if r.Password, err = s.hasher.Hash(r.Password); err != nil {
+		return nil, err
+	}
+
+	// Make sure that username is unique
+	found, err := s.Contains("username", r.Username)
+
+	if err != nil || found {
+		return nil, errors.New("Username not unique")
+	}
+
+	acc := account.DefaultAccount{
+		ID:       r.ID,
+		Username: r.Username,
+		Password: r.Password,
+		Data:     r.Data,
+	}
+
+	res, err := rdb.Table(accountsTable).Insert(acc).RunWrite(s.session)
+
+	if err != nil {
+		return nil, errors.New(err)
+	} else if res.Errors > 0 {
+		return nil, errors.New(res.FirstError)
+	}
+
+	return &acc, nil
+}
+
+func (s *Store) Contains(field string, value string) (bool, error) {
+	res, err := rdb.Table(accountsTable).Field(field).Contains(value).Run(s.session)
+	if err != nil {
+		return false, err
+	}
+	defer res.Close()
+
+	var found bool
+	err = res.One(&found)
+
+	if err != nil {
+		return false, err
+	}
+
+	return found, nil
+}
+
+func (s *Store) Get(id string) (account.Account, error) {
+	// Query account
+	result, err := rdb.Table(accountsTable).Get(id).Run(s.session)
+	defer result.Close()
```

Comment:
This should not be a problem, but for reliabilities sake, i'll add error checks before to cover all possible cases.